app/src/main/java/com/example/tasky/agenda/data/AgendaRepositoryImpl.kt

```diff
@@ -0,0 +1,73 @@
+package com.example.tasky.agenda.data
+
+import android.annotation.SuppressLint
+import com.example.tasky.agenda.data.local.AgendaDao
+import com.example.tasky.agenda.data.mapper.toDomain
+import com.example.tasky.agenda.data.mapper.toEntity
+import com.example.tasky.agenda.data.remote.AgendaService
+import com.example.tasky.agenda.data.util.asTodayEndLong
+import com.example.tasky.agenda.data.util.asTodayStartLong
+import com.example.tasky.agenda.data.util.toCurrentTimeMilli
+import com.example.tasky.agenda.domain.AddNeedleToAgenda
+import com.example.tasky.agenda.domain.AgendaRepository
+import com.example.tasky.agenda.domain.model.AgendaItem
+import com.example.tasky.core.data.Resource
+import com.example.tasky.core.data.remote.safeApiCall
+import java.time.LocalDate
+import javax.inject.Inject
+
+@SuppressLint("NewApi")
+class AgendaRepositoryImpl @Inject constructor(
+    private val agendaService: AgendaService,
+    private val agendaDao: AgendaDao
+) : AgendaRepository {
+
+    override suspend fun getAgenda(date: LocalDate): Resource<List<AgendaItem>> {
+        val result = safeApiCall {
+            agendaService.getAgenda(time = date.toCurrentTimeMilli())
+        }
+
+        when (result) {
+            is Resource.Success -> {
+                with(result.data) {
+                    events.map { agendaDao.insertEvent(it.toEntity()) }
+                    reminders.map { agendaDao.insertReminder(it.toEntity()) }
+                    tasks.map { agendaDao.insertTask(it.toEntity()) }
+                }
+                val localEvents = getLocalEventsByDate(date)
+                val localReminders = getLocalRemindersByDate(date)
+                val localTasks = getLocalTasksByDate(date)
+
+                val agendaItems: MutableList<AgendaItem> =
+                    (localEvents + localReminders + localTasks).toMutableList()
+
+                agendaItems.sortBy { it.time }
+
+                AddNeedleToAgenda(date.toEpochDay(), agendaItems)
+
+                return Resource.Success(agendaItems)
+            }
+
+            is Resource.Error -> return Resource.Error(result.errorMessage)
```

Comment:
Here the logic is a bit weird:

You make an api call
If fails, then you return an error
Even if the DB has items

You should return items. Since the app is Offline-First, you want it to still work even if there´s no internet connection


What I recommend, is the following (@philipplackner might have a different answer)

- Have the method return a Flow<List<AgendaItem>>
- Get the items from DB, and emit them
- Make API call, and save them to db
- Get the items from DB again (since you need to grab the updated ones after API call) and emit those - I think this isn´t fully if your Room db returns a Flow

Reply:
Yes or alternatively this:
1. Have a function that just reads all agenda items for the day from the DB in form of a Flow
2. The UI observes that function and just shows the items from that flow
3. Whenever you refresh/sync, your repo just inserts the new items in the DB table which will automatically trigger a refresh of the UI since it listens to these DB updates